### PR TITLE
[LLM Runtime] Add Script for PPL Evaluation

### PR DESCRIPTION
**Pull request by DDEle**
## Type of Change: Feature
API not changed

## Description
- Add PPL script
- Add `cal_diff.py` script for future CI @zhenwei-intel
- Fix UT (`test_llm_runtime.py`)

## Expected Behavior & Potential Risk
N/A

## How has this PR been tested?
Internal extension test:
- Llama2: https://inteltf-jenk.sh.intel.com/job/ITREX-cpp-graph-extension-WIP/32/artifact/report.html
- GPT-NEOX and OPT: https://inteltf-jenk.sh.intel.com/job/ITREX-cpp-graph-extension-WIP/33/artifact/report.html
- All models: https://inteltf-jenk.sh.intel.com/job/ITREX-cpp-graph-extension-WIP/34/

## Dependency Change?
No

### CODE REVIEW

File: intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp [hunk 1]
```diff
@@ -403,7 +441,7 @@ model_token Model::post_process(float* logits) {
 
 int Model::quant_model(const std::string& model_path, const std::string& out_path, const std::string& weight_dtype,
                        const std::string& alg, int group_size, const std::string& scale_dtype,
-                       const std::string& compute_dtype, bool use_ggml) {
+                       const std::string& compute_dtype, bool use_ggml, int threads) {
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp:444]: for clients?

File: intel_extension_for_transformers/llm/runtime/graph/scripts/perplexity.py [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+    )
+    group.add_argument("--memory_dtype",
+                       type=str,
+                       help="Data type of the kv memory",
+                       choices=['f32', 'f16', 'auto'],
+                       default="auto")
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="Evaluate perplexity for a model givan a dataset")
```
Comment (airMeng) [inline intel_extension_for_transformers/llm/runtime/graph/scripts/perplexity.py:240]: can we reuse the run.py?

**Review by airMeng — COMMENTED** [review]
add a note on homepage readme

File: intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp (hunk 1 above)
Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp:444]: There is a `int threads` in `generate` and `generate_tokens` and I guess they are no more "for clients" than `quant_model`.

File: intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp [hunk 3]
```diff
@@ -149,56 +166,64 @@ void Model::reinit() {
...
+  const auto& input_id0 = input_ids[0];  // currently only support single batch
+  if (input_id0.empty()) {               // use internel input id
+    if (curr_input_ids.empty()) {
+      fprintf(stderr, "%s: error: no input\n", __func__);
+      return empty_ret;
     }
+  } else if (!curr_input_ids.empty()) {
+    fprintf(stderr, "%s: error: prompt confliction\n", __func__);
+    return empty_ret;
+  } else if (input_id0.size() > n_ctx - 4) {  // long input_id0 and empty curr_input_ids
```
Comment (zhentaoyu) [inline intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp:214]: Do you think we can move this checking outside from this `evaluate_` function? I also want to use it before beam search generation. Or move this `evaluate_` into `model_utils.cpp`?

File: intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp [hunk 4]
```diff
@@ -58,14 +58,23 @@ class Model {
   void init_model(const std::string& model_path, int n_predict, int n_batch, int ctx_size, int seed, int threads,
                   float repetition_penalty, int num_beams, bool do_sample, int top_k, float top_p, float temperature,
                   int min_new_tokens, float length_penalty, bool early_stopping, int n_keep, int n_discard,
-                  bool shift_roped_k, int batch_size, model_vocab::id pad_token);
+                  bool shift_roped_k, int batch_size, model_vocab::id pad_token, const std::string& memory_dtype);
   void reinit();
   std::vector<std::vector<model_token>> generate(const std::vector<std::vector<model_token>>& input_ids);
   std::vector<std::vector<model_token>> generate_tokens(const std::vector<std::vector<model_token>>& input_ids);
+  const std::vector<float>& evaluate_(const std::vector<std::vector<model_token>>& input_ids);
+  std::vector<std::vector<float>> evaluate(const std::vector<std::vector<model_token>>& input_ids) {
```
Comment (zhenwei-intel) [inline intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp:66]: how about returning numpy for better performance

File: intel_extension_for_transformers/llm/runtime/graph/scripts/perplexity.py [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+    parser.add_argument('--model_name', type=str, default="~/Llama-2-7b-chat-hf")
+    parser.add_argument('--dataset_name', type=str, default="~/pg19-data-test")
+    parser.add_argument('--ctx_size', type=int, default=256)
+    parser.add_argument('--prompt_size', type=int)
+    parser.add_argument('--n_threads', type=int)
+    parser.add_argument('--n_pred_per_sample', type=int)
+    parser.add_argument('--n_sampels', type=int)
+    parser.add_argument('--data_text_concat', action="store_true")
+    parser.add_argument('--fig_path', type=str, default="out/ppl.png")
+    add_quant_args(parser)
```
[the remark this answers is not in the document]
Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/scripts/perplexity.py:250]: I'm considering add direct weight access API and fp32 test can be based on that

**Comment by zhenwei-intel** [comment]
Do we have PPL data of pytorch and that of cpp graph fp32/int4?

File: intel_extension_for_transformers/llm/runtime/graph/scripts/perplexity.py (hunk 2 above)
Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/scripts/perplexity.py:240]: I tried but failed. There are some args in run.py (e.g. --color --prompt --build_dir) that does not make sense to ppl while there are some args (e.g. --dataset_name --n_pred_per_sample  --data_text_concat --fig_path) for ppl does not make sense to run.py.

File: intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp (hunk 4 above)
Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp:66]: Great idea. 99%+ users will cast the result to numpy/torch for data procession. It also make possible to do zero-cost reshaping, while will be useful if we support multi-batch logits output in the future.

> https://stackoverflow.com/a/48483811/9817693
> > Pytorch docs:
> > 
> > If a numpy.ndarray, torch.Tensor, or torch.Storage is given, a new tensor that shares the same data is returned. If a Python sequence is given, a new tensor is created from a copy of the sequence.

File: intel_extension_for_transformers/llm/runtime/graph/scripts/perplexity.py (hunk 2 above)
Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/scripts/perplexity.py:240]: It is a good idea for me. But I guess there is enough python changes for this PR. Will try to generalize args in run/convert/inference/ppl to a `graph/scripts/utils.py` file

File: intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp (hunk 4 above)
Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp]: Updated
https://github.com/intel/intel-extension-for-transformers/blob/856b6cd9a05a43ed5dc7c0f40fe256cec6ad5324/intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp#L71-L79

File: intel_extension_for_transformers/llm/runtime/graph/scripts/perplexity.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/scripts/perplexity.py:258]: Added the `quantized_weight_path` argument
https://github.com/intel/intel-extension-for-transformers/blob/856b6cd9a05a43ed5dc7c0f40fe256cec6ad5324/intel_extension_for_transformers/llm/runtime/graph/scripts/perplexity.py#L97-L110

File: intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp (hunk 3 above)
Comment (zhentaoyu) [reply intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp:220]: will be resolved after detailed discussion.

File: intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp [hunk 6]
```diff
@@ -149,56 +172,64 @@ void Model::reinit() {
...
-  if (curr_input_ids.empty()) {
-    if (input_ids[0].size() > n_ctx - 4) {
-      fprintf(stderr, "\n%s: Warning: prompt is too long (%d tokens, max %d), will be truncated\n", __func__,
-              input_ids[0].size(), n_ctx - 4);
-      curr_input_ids.resize(n_ctx - 4);
-      std::copy(input_ids[0].end() - n_ctx - 4, input_ids[0].end(), curr_input_ids.begin());
-    } else {
-      curr_input_ids = input_ids[0];
+
+  const auto& input_id0 = input_ids[0];  // currently only support single batch
```
Comment (zhenwei-intel) [inline intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp:211]: error here, input_list is empty when next generate
```
        while True:
            response = self.model.generate(input_ids=input_list)
            input_list = []  # next-token stage will use previous output
```

Comment (DDEle) [reply intel_extension_for_transformers/llm/runtime/graph/application/main_pybind.cpp:211]: It is by design. To be discussed

**Comment by DDEle** [comment]
> Do we have PPL data of pytorch and that of cpp graph fp32/int4?

image
https://inteltf-jenk.sh.intel.com/job/ITREX-cpp-graph-extension-WIP/32/artifact/report.html

**Review by zhenwei-intel — APPROVED** [review]

**Comment by DDEle** [comment]
gpt-neox-20b and opt-1.3b checked in this internal extension test: https://inteltf-jenk.sh.intel.com/job/ITREX-cpp-graph-extension-WIP/33/artifact/report.html

**Review by a32543254 — APPROVED** [review]
LGTM

**Review by airMeng — APPROVED** [review]

**Review by hshen14 — COMMENTED** [review]
What's the purpose to introduce ppl script here? Can we just reuse lm-eval-harness to measure the ppl?

**Comment by DDEle** [comment]
> What's the purpose to introduce ppl script here? Can we just reuse lm-eval-harness to measure the ppl?

It is for internal usage of LLM Runtime to keep track of the inference accuracy with minimal test time addition. It also works as a usage example of the LLM Runtime (llama.cpp also has a ppl script in its `examples` directory). We will investigate lm-eval-harness and see if it can help.
